PHP 8.0 support

jfernancordova opened the pull request:
**_Issue_** #74 

|    Q          |   A
|-------------- | ------
| New Feature   | yes

### Description

- [x] Modify composer.json to provide support for PHP 8.0 by adding the constraint ~8.0.0
- [x]  Modify composer.json to drop support for PHP less than 7.3
- [x] Modify composer.json to implement phpunit 9.3 which supports PHP 7.3+
- [x] Modify .travis.yml to ignore platform requirements when installing composer dependencies (simply add --ignore-platform-reqs to COMPOSER_ARGS env variable)
- [x] Modify .travis.yml to add PHP 8.0 to the matrix (NOTE: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [x] Modify source code in case there are incompatibilities with PHP 8.0

boesing commented on composer.json:
```diff
@@ -25,7 +25,7 @@
     },
     "require": {
-        "php": "^7.1",
+        "php": "~8.0.0",
```
Please use `^7.3 || ~8.0.0` as descripted in #74

jfernancordova replied:
Fixed

boesing commented on composer.json:
```diff
@@ -42,7 +42,7 @@
         "laminas/laminas-session": "^2.8",
         "laminas/laminas-uri": "^2.5",
-        "phpunit/phpunit": "^7.5.20 || ^8.5.2",
+        "phpunit/phpunit": "^9.3.*",
```
Its okay to use `^9.3` here

jfernancordova replied:
Fixed

boesing commented on .travis.yml:
```diff
@@ -6,7 +6,7 @@ cache:
 env:
   global:
-    - COMPOSER_ARGS="--no-interaction"
+    - COMPOSER_ARGS="--ignore-platform-reqs --no-interaction"
```
please only ignore platform requirements for nightly (which is not yet part of this file)

jfernancordova replied:
Ok, I get it, I'll revert this

jfernancordova commented:
@boesing I just added PHP 8.0 to the matrix in travis.yml, anything else?

boesing commented:
> @boesing I just added PHP 8.0 to the matrix in travis.yml, anything else?

You have to use `nightly` in travis as there is no `8.0` version available yet.
After this, you can convert your PR to `Ready for review` by yourself and then we'll see what travis says 👍

jfernancordova commented:
Perfect

boesing commented:
@jfernancordova Please check mezzio/mezzio-cors [.travis.yml](https://github.com/mezzio/mezzio-cors/blob/1.1.x/.travis.yml) to see how you have to change the `.travis.yml` to finish this PR. Just ignore the `allow_failures` part as this is fixed in a PR in that repository.

If travis is "green", this is done 🎉

boesing reviewed (changes requested):
Please remove PHP <7.3 from travis aswell.

jfernancordova commented:
@boesing it's failing for **_laminas/laminas-uri_**, do you think that upgrading this dependency, could it solve this error in travis?

glensc commented:
`--ignore-platform-reqs` should never be used. it does not just ignore `ext-foo` requirement, but also PHP version dependency installing packages that are not compatible at all with the runtime. for example running in 5.6 installing package resolving that requires PHP 7.1 and then execution results fatal errors due invalid syntax. 

if you don't have ext-foo installed, fake it with local `composer config platform.ext-foo '0'` instead.

example:
- https://github.com/eventum/eventum/blob/v3.9.0/bin/releng/configure.sh#L11-L13

boesing commented:
> `--ignore-platform-reqs` should never be used. it does not just ignore `ext-foo` requirement, but also PHP version dependency installing packages that are not compatible at all with the runtime. for example running in 5.6 installing package resolving that requires PHP 7.1 and then execution results fatal errors due invalid syntax.
> 
> if you don't have ext-foo installed, fake it with local `composer config platform.ext-foo '0'` instead.
 
Thanks for the information.
You are right, its not a good solution. However, this is the only way to initially provide PHP 8.0 support on packages. Since we are not using `composer v2` which wont happen in the near future anyways, there is no way to use `--ignore-platform-req=php` and thus, we *have* to use `--ignore-platform-reqs`. Without ignoring platform reqs, this package is not testable by travis.
However, this should not be used in upstream projects but for travis this is absolutely fine.

boesing commented on .travis.yml:
```diff
@@ -6,28 +6,13 @@ cache:
 env:
   global:
-    - COMPOSER_ARGS="--no-interaction"
+    - COMPOSER_ARGS="--no-interaction --ignore-platform-reqs"
```
Please only ignore platform reqs for nightly builds

jfernancordova replied:
Fixed

glensc commented:
> we have to use `--ignore-platform-reqs`. 

that is incorrect. what is the error you trying to avoid with the flag?

boesing commented:
> > we have to use `--ignore-platform-reqs`.
> 
> that is incorrect. what is the error you trying to avoid with the flag?

I want to avoid having trouble with 3rd-party composer dependencies which are not yet PHP 8.0 compatible as no dependency wont install when one of them are not compatible (yet).

Lets take `laminas/laminas-uri` for example. You wont be able to use `composer install` unless you either provide `--ignore-platform-reqs` (to ignore that the package is not yet compatible with PHP 8.0) or by using `--no-dev`.

Feel free to resolve dependency tree on PHP 8.0 without passing `--ignore-platform-reqs`.

But as you wanted to have the error:

```
PATH="/usr/local/Cellar/php@8.0/8.0.0/bin:$PATH" php80 /usr/local/bin/composer install
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for laminas/laminas-config ^2.6 -> satisfiable by laminas/laminas-config[2.6.0].
    - laminas/laminas-config 2.6.0 requires php ^5.5 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
  Problem 2
    - laminas/laminas-cache 2.9.0 requires php ^5.6 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - laminas/laminas-cache 2.8.3 requires php ^5.6 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - laminas/laminas-cache 2.8.2 requires php ^5.6 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - laminas/laminas-cache 2.8.1 requires php ^5.6 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - laminas/laminas-cache 2.8.0 requires php ^5.6 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - laminas/laminas-cache 2.7.2 requires php ^5.5 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - laminas/laminas-cache 2.7.1 requires php ^5.5 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - laminas/laminas-cache 2.7.0 requires php ^5.5 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - laminas/laminas-cache 2.6.1 requires php ^5.5 || ^7.0 -> your PHP version (8.0.0-dev) does not satisfy that requirement.
    - Installation request for laminas/laminas-cache ^2.6.1 -> satisfiable by laminas/laminas-cache[2.6.1, 2.7.0, 2.7.1, 2.7.2, 2.8.0, 2.8.1, 2.8.2, 2.8.3, 2.9.0].
```

boesing commented:
@jfernancordova As you can see, the tests are still failing. This is due to the fact that there are some strange circulare dependencies going on in this package.

Please run `composer test` locally. You will see several problems which needs to get fixed (as there are missing methods such as `assertAttributeSame` and deprecations for prophecy).

You can use `phpspec/prophecy-phpunit` as a dev dependency and add the trait to the tests which are using the `prophesize` method.
To resolve dependencies, you can temporarily add `"version": "2.14.0",` to the `composer.json` (but please do not commit that as this is only needed locally). When you've added `phpspec/prophecy-phpunit` you can remove the `version` property from `composer.json` as the dependency tree should be resolved locally.

@Ocramius @weierophinney Any ideas how to solve such circular dependencies in the future? Is there a way of adding the `version` property to `composer.json` from within the `automatic-releases`?

jfernancordova commented:
> @jfernancordova As you can see, the tests are still failing. This is due to the fact that there are some strange circulare dependencies going on in this package.
> 
> Please run `composer test` locally. You will see several problems which needs to get fixed (as there are missing methods such as `assertAttributeSame` and deprecations for prophecy).
> 
> You can use `phpspec/prophecy-phpunit` as a dev dependency and add the trait to the tests which are using the `prophesize` method.
> To resolve dependencies, you can temporarily add `"version": "2.14.0",` to the `composer.json` (but please do not commit that as this is only needed locally). When you've added `phpspec/prophecy-phpunit` you can remove the `version` property from `composer.json` as the dependency tree should be resolved locally.
> 
> @Ocramius @weierophinney Any ideas how to solve such circular dependencies in the future? Is there a way of adding the `version` property to `composer.json` from within the `automatic-releases`?

Ok. let me check.

jfernancordova commented:
@boesing, I fixed all tests and executed `composer test` locally, however, I'm seeing the Build from travis CI, and other dependencies such as: laminas/laminas-http are not working with laminas/laminas-validator, can you suggest any solution for this?, thx

boesing commented:
@jfernancordova You can add `"version": "2.14.0"` to `composer.json` in a single commit. 
After that, push to your branch. Let travis do its job. 
- When everything is green, *hooray*. Copy the link to the test-results and paste them in here.
- If not, try to determine the problems and fix them and post the link to the green test-results here.

When done, drop the commit from history and force-push back to your branch 👍 

Not sure if there is a better way but thats what comes to my mind.

jfernancordova commented:
> @jfernancordova You can add `"version": "2.14.0"` to `composer.json` in a single commit.
> After that, push to your branch. Let travis do its job.
> 
>     * When everything is green, _hooray_. Copy the link to the test-results and paste them in here.
> 
>     * If not, try to determine the problems and fix them and post the link to the green test-results here.
> 
> 
> When done, drop the commit from history and force-push back to your branch +1
> 
> Not sure if there is a better way but thats what comes to my mind.

Hi @boesing, this is the link: https://travis-ci.com/github/laminas/laminas-validator/builds/190545033, I hope it's ok.

jfernancordova commented:
@boesing what do you think?, I just changed some code and fixed all tests, anything else?

froschdesign commented on test/Db/NoRecordExistsTest.php:
```diff
@@ -204,6 +204,14 @@ public function testThrowsExceptionWithNoAdapter()
      */
     public function testWithSchema()
     {
+        if (explode('.', PHP_VERSION)[0] >= 8.0) {
```
[PHP's `version_compare`](https://www.php.net/manual/function.version-compare.php) can be used here.

jfernancordova replied:
Great, I'll fix these conditionals, thx.

glensc replied:
if you are comparing the php version, use PHP_VERSION_ID constant

glensc replied:
altho, you should use the requires annotation here instead:


```
     * @requires PHP < 8.0
```

- https://phpunit.readthedocs.io/en/9.3/incomplete-and-skipped-tests.html#skipping-tests-using-requires

jfernancordova replied:
Great, thank you @glensc!

froschdesign commented on test/Db/NoRecordExistsTest.php:
```diff
@@ -218,6 +226,14 @@ public function testWithSchema()
      */
     public function testWithSchemaNoResult()
     {
+        if (explode('.', PHP_VERSION)[0] >= 8.0) {
```
The same here and so on…

froschdesign commented on test/InArrayTest.php:
```diff
@@ -142,10 +142,10 @@ public function testStrictComparisons()
         $this->assertFalse($validator->isValid('b'));
         $this->assertFalse($validator->isValid('a'));
-        $this->assertTrue($validator->isValid('A'));
+        $this->assertFalse($validator->isValid('A'));
```
This means a change of logic. This cannot be correct and should be checked again.

jfernancordova replied:
You're right, I will check this.

froschdesign commented on test/InArrayTest.php:
```diff
@@ -142,10 +142,10 @@ public function testStrictComparisons()
         $this->assertFalse($validator->isValid('0'));
         $this->assertFalse($validator->isValid('1a'));
-        $this->assertTrue($validator->isValid(0));
+        $this->assertFalse($validator->isValid(0));
```
The same here. The haystack contains the following values:

https://github.com/laminas/laminas-validator/blob/83c2a967cd823a76a81589e895363aac73b4d646/test/InArrayTest.php#L133-L137

froschdesign commented on test/ValidatorPluginManagerFactoryTest.php:
```diff
@@ -28,10 +31,7 @@ public function testFactoryReturnsPluginManager()
             // laminas-servicemanager v3
-            $this->assertAttributeSame($container, 'creationContext', $validators);
-        } else {
-            // laminas-servicemanager v2
-            $this->assertSame($container, $validators->getServiceLocator());
+            $this->assertObjectHasAttribute('creationContext', $validators);
```
This section is no longer useful and testing internal implementation details is the wrong way.

jfernancordova replied:
Ok, I'll remove the condition

froschdesign commented on src/InArray.php:
```diff
@@ -193,8 +193,16 @@ public function isValid($value)
                     }
-
-                    if ($el == $value) {
+    
+                    if ($el == $value && self::COMPARE_STRICT != $this->strict) {
+                        if (InArray::COMPARE_NOT_STRICT_AND_PREVENT_STR_TO_INT_VULNERABILITY == $this->strict) {
+                            if ($this->inArrayRecursive($el, $haystack)) {
+                                return true;
+                            }
+                            return false;
+                        }
+                        return true;
+                    }elseif (InArray::COMPARE_NOT_STRICT == $this->strict){
```
If you find a problem in this the validator independent of the PHP 8 support then create a separate pull request with an unit test. Otherwise, please revert the changes in this file.  Thanks!

jfernancordova replied:
Ok, I revert this change and add a version compare in the test, because only 2 or 3 cases are not working with PHP 8.0

froschdesign replied:
@jfernancordova
> …add a version compare in the test, because only 2 or 3 cases are not working with PHP 8.0.

I am afraid there is a misunderstanding here. Sorry!
All cases must be passed otherwise it would be a break of the behaviour. The validator must produce the same results in all PHP versions.

jfernancordova replied:
I see, @froschdesign so if I revert these changes, the tests aren't going to work correctly. I'll try to fix this in this PR first.

froschdesign replied:
Thanks in advance! 👍 👍 👍

jfernancordova commented:
@froschdesign I analyzed the code again. I could understand the complexity, so I did a refactor in which the code is cleaner now. What do you think?. Any suggestions are welcome.

boesing commented on src/AbstractValidator.php:
```diff
@@ -132,11 +132,11 @@ public function setOptions($options = [])
             $fname = 'set' . ucfirst($name);
             $fname2 = 'is' . ucfirst($name);
-            if (($name !== 'setOptions') && method_exists($this, $name)) {
+            if (($name !== 'setOptions') && get_parent_class($this) && method_exists($this, $name)) {
```
Could you explain why you've added these additional checks with `get_parent_class`?
As this is an abstract class, this check is *always* true or do I miss something here?

In case we do not need this additional check (and I think so), please revert this change.

jfernancordova replied:
Yes, I just considered an extra validation, but you're right, I'm going to revert this change.

boesing commented on src/InArray.php:
```diff
@@ -216,14 +220,32 @@ public function isValid($value)
+    
+    private function isMultiArray($array)
+    {
+        $rv = array_filter($array, 'is_array');
```
depending on the size of the given array, you might want to use `foreach` with early return instead. This will avoid unnecessary CPU cycles.

boesing commented on src/InArray.php:
```diff
@@ -216,14 +220,32 @@ public function isValid($value)
         return false;
     }
+    
+    private function isMultiArray($array)
```
Please also add proper typehints to the methods parameter and return value.

boesing commented on .travis.yml:
```diff
@@ -13,21 +13,6 @@ env:
-    - php: 7.1
-      env:
-        - DEPS=latest
-        - EXECUTE_HOSTNAME_CHECK=true
```
Oh, as far as I can see here, we dropped some checks.
Could you re-add these to PHP 7.4?
Sorry, realized just yet.
After this, we should be good to go (I think the CS check will find some problems, maybe you check yourself by locally executing `composer cs-check`).

jfernancordova replied:
No problem!

boesing commented on .travis.yml:
```diff
@@ -40,6 +25,15 @@ matrix:
     - php: 7.4
       env:
         - DEPS=latest
+        - EXECUTE_HOSTNAME_CHECK=true
```
`TEST_COVERAGE` and `CS_CHECK` are missing in here aswell.

jfernancordova replied:
Uh, I'll add them too, sorry.

jfernancordova replied:
@boesing Adding TEST_COVERAGE is returning an error: PHP Fatal error: Uncaught Error: Maximum function nesting level of '256' reached, aborting! in /home/travis/build/laminas/laminas-validator/vendor/nikic/php-parser/lib/PhpParser/NodeAbstract.php. I think it's about of php.ini or xdebug.ini configuration file. I modified the travis yml, on the before_install section `echo "xdebug.max_nesting_level=3000" >> xdebug.ini`

jfernancordova commented:
@boesing Adding `TEST_COVERAGE` is returning an error like: `PHP Fatal error:  Uncaught Error: Maximum function nesting level of '256' reached, aborting! in /home/travis/build/laminas/laminas-validator/vendor/nikic/php-parser/lib/PhpParser/NodeAbstract.php`. I think it's about of `php.ini` configuration file.

boesing commented on .travis.yml:
```diff
@@ -40,8 +25,21 @@ matrix:
 
 before_install:
+  - export INI=~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
+  - echo xdebug.max_nesting_level = 3000 >> $INI
```
Does the error for which this change is necessary happen for coverage only? 
I'd prefer having this within `phpunit.xml.dist` instead as its phpunit related and could be a problem in local development aswell.

See phpunit configuration documentation: https://phpunit.readthedocs.io/en/9.3/configuration.html#the-ini-element

jfernancordova replied:
Yes, only for coverage, ok, I'll check about it.

boesing commented on composer.json:
```diff
@@ -2,6 +2,7 @@
     "name": "laminas/laminas-validator",
     "description": "Validation classes for a wide range of domains, and the ability to chain validators to create complex validation criteria",
     "license": "BSD-3-Clause",
+    "version": "2.14.0",
```
after tests are green, please remove this from the `composer.json` by removing the related commit.

jfernancordova replied:
Ok

jfernancordova commented:
@boesing Before removing the composer's version, all tests are green!
https://travis-ci.com/github/laminas/laminas-validator/builds/197589684

boesing reviewed (approved):
LGTM

samsonasik commented:
please update laminas-http require-dev to ^2.14

jfernancordova commented:
> please update laminas-http require-dev to ^2.14

Ok!

Ocramius commented on src/InArray.php:
```diff
@@ -213,17 +217,35 @@ public function isValid($value)
                     }
                 }
+
+                if (! $this->isMultiArray($haystack) && array_key_exists($value, array_flip($haystack))) {
```
Is flipping safe here? What of the array contains objects?

jfernancordova replied:
Hi @Ocramius, it was the solution I thought could work and all tests passed: https://travis-ci.com/github/laminas/laminas-validator/builds/197589684, unfortunately, the last implementation breaks with PHP 8, if you have any suggestions, let me know, please!

However, I could reevaluate it!

jfernancordova replied:
I'm working on it!

jfernancordova replied:
@Ocramius what do you think about this?
```
$haystack = is_object($haystack) ? get_object_vars($haystack) : $haystack;
if (! $this->isMultidimensionalArray($haystack) && array_key_exists($value, array_flip($haystack))) {
    return true;
}
```

Ocramius commented on src/InArray.php:
```diff
@@ -213,17 +217,35 @@ public function isValid($value)
+                }
+            }
+
+            if (array_search($value, $haystack, self::COMPARE_STRICT == $this->strict)) {
```
Can `==` and `!=` be replaced with strict comparison (and is it safe to do so)?

jfernancordova replied:
Sorry @Ocramius, I don't understand you, do you want to replace it with strict comparison or know if it safe?

glensc replied:
@jfernancordova, I read it as to make a change to do strict compare but validate first is the change legal first, i.e not introducing a bug by making such change.

Ocramius commented on src/InArray.php:
```diff
@@ -213,17 +217,35 @@ public function isValid($value)
         return false;
     }
+
+    private function isMultiArray(array $array): bool
```
s/multi/multidimensional?

jfernancordova replied:
yes multidimensional!

glensc replied:
@richardjennings @Ocramius suggested here you to change the word `multi` with `multidimensional`

jfernancordova replied:
I get it now, thanks @glensc!

Ocramius commented on test/Db/NoRecordExistsTest.php:
```diff
@@ -200,6 +200,10 @@ public function testThrowsExceptionWithNoAdapter()
      * Test that schemas are supported and run without error
      *
+     * @requires PHP < 8.0
+     * For PHP 8.0 Library laminas/laminas-stdlib/src/ArrayUtils.php needs to be upgraded
```
This means that we aren't yet php8-compatible, and we should fix that dependency (bug) first

jfernancordova replied:
That's right!

Ocramius commented on test/Db/NoRecordExistsTest.php:
```diff
@@ -214,6 +218,10 @@ public function testWithSchema()
     /**
      * Test that schemas are supported and run without error
      *
+     * @requires PHP < 8.0
```
This means that we aren't yet php8-compatible, and we should fix that dependency (bug) first

jfernancordova replied:
That's right!

Ocramius reviewed (changes requested):
We really can't use `@requires php<8` for releasing this: if it doesn't run, it isn't ready 😞

froschdesign commented on .travis.yml:
```diff
@@ -40,6 +25,17 @@ matrix:
+        - EXECUTE_HOSTNAME_CHECK=true
+        - CS_CHECK=true
+        - TEST_COVERAGE=true
+    - php: nightly
```
Version 8.0 should be used here.

Ocramius commented on .travis.yml:
```diff
@@ -13,33 +13,33 @@ env:
         - DEPS=lowest
-    - php: 7.4
+        - COMPOSER_ARGS="--ignore-platform-reqs"
+    - php: nightly
```
If we test nightly, we should probably allow failures there (it's very unstable)

jfernancordova replied:
so I remove nightly, right?

samsonasik replied:
iirc, nightly was for php 8 dev, I think it can be removed.

jfernancordova replied:
OK!

Ocramius commented on test/ValidatorPluginManagerFactoryTest.php:
```diff
@@ -15,24 +15,18 @@
         $validators = $factory($container, ValidatorPluginManagerFactory::class);
         $this->assertInstanceOf(ValidatorPluginManager::class, $validators);
-
-        if (method_exists($validators, 'configure')) {
```
Ìsn't v3 "current"?

jfernancordova replied:
@Ocramius I tried to set these asserts, especially v3, and they're deprecated, that's why I remove this block.

lahivee commented:
Hey guys, is there any way to push this through donations? We are depending on this package to support PHP8 for this package: https://github.com/beyondcode/expose/pull/163

danmichaelo commented:
Per https://github.com/laminas/laminas-validator/pull/79#discussion_r550776411 , can we remove `laminas/laminas-cache` as a dev dependency? Couldn't find a single reference to caches in this repo: `find . -iname '*.php' -exec grep -i 'Cache' {} \;`

weierophinney commented on composer.json:
```diff
@@ -36,13 +36,14 @@
         "laminas/laminas-db": "^2.7",
         "laminas/laminas-filter": "^2.6",
-        "laminas/laminas-http": "^2.5.4",
+        "laminas/laminas-http": "^2.14",
```
Bump to 2.14.2, please

weierophinney commented on src/InArray.php:
```diff
@@ -173,7 +173,7 @@ public function isValid($value)
             && (is_int($value) || is_float($value))) {
-            $value = (string) $value;
+            $value = (string)$value;
         }
```
These changes are all unrelated... and I suspect you ran a CS tool over this other than our own `composer cs-check` and `composer cs-fix` (which use phpcs and the laminas-coding-standard). Please either revert these changes, or run `composer cs-fix` and commit the updates.

jfernancordova replied:
Ok!

weierophinney commented on src/InArray.php:
```diff
@@ -191,10 +191,14 @@ public function isValid($value)
+                        return true;
+                    }
+
+                    if (self::COMPARE_NOT_STRICT == $this->strict) {
```
What is this change for, exactly? It seems unrelated to the bump to PHP 8...

weierophinney commented on src/InArray.php:
```diff
@@ -213,17 +217,45 @@ public function isValid($value)
             }
 
-            if (in_array($value, $haystack, self::COMPARE_STRICT == $this->strict)) {
+            if (array_search($value, $haystack, self::COMPARE_STRICT == $this->strict)) {
```
Please revert this change; they are not semantically the same!

We want to see if the value is in the array, not return the first match. `in_array()` is the correct one to use here.

weierophinney commented on src/InArray.php:
```diff
@@ -213,17 +217,45 @@ public function isValid($value)
+        return false;
+    }
+
+    private function isArrayKeyExists(string $key, array $haystack): bool
```
This is misnamed, and duplicates the functionality of `in_array()`. Please remove, and replace calls to this method with `in_array($value, $haystack)`.

weierophinney commented on src/InArray.php:
```diff
@@ -213,17 +217,45 @@ public function isValid($value)
                     }
                 }
+
+                if (! $this->isMultidimensionalArray($haystack) && $this->isArrayKeyExists($value, $haystack)) {
```
This is unrelated to updating to PHP 8. Please remove.

weierophinney reviewed (changes requested):
There are a lot of CS changes that are (a) unrelated to updating to PHP 8, and (b) incorrect by our coding standards. Additionally, the InArray validator has a number of changes outside the scope of the PR (adding PHP 8 support), and without new tests to validate the new functionality.

Please run `composer cs-fix`, then `composer cs-check`, fixing and committing any changes needed to make the checks pass. Also, please back out the changes in the `InArray` validator; if you want to see them in a future version, please submit them as a separate PR, with tests demonstrating and validating the changes.

jfernancordova commented:
Hi @weierophinney, As you can see, reverting the changes from inArray.php class, break the tests with PHP 8.0, https://travis-ci.com/github/laminas/laminas-validator/jobs/468944524, that's why I changed the code, the native function in_array seems not working as the previous versions.

However, I'm going to create another PR with the new changes and their tests for inArray validator.

jfernancordova committed:
Upgrade PHP 8.0

Signed-off-by: José Fernando Cordova <jfernancordova@gmail.com>

lahivee commented:
@jfernancordova thank you so much for your work on this!

weierophinney commented:
@jfernancordova —

> reverting the changes from inArray.php class, break the tests with PHP 8.0, https://travis-ci.com/github/laminas/laminas-validator/jobs/468944524, that's why I changed the code, the native function in_array seems not working as the previous versions.
> 
> However, I'm going to create another PR with the new changes and their tests for inArray validator.

I'm fine with including the changes here; in fact, since the PHP 8 tests _fail_ without them, then we need them before merging. I just didn't understand why you weren't using `in_array()` directly, and there were no comments in the code to indicate why.

Speaking of: it may be good to create a bug on bugs.php.net about this. The PHP 8 changelog does not mention anything about the change in behavior, and I don't see any bugs currently filed around it.

jfernancordova commented:
> @jfernancordova —
> 
> > reverting the changes from inArray.php class, break the tests with PHP 8.0, https://travis-ci.com/github/laminas/laminas-validator/jobs/468944524, that's why I changed the code, the native function in_array seems not working as the previous versions.
> > However, I'm going to create another PR with the new changes and their tests for inArray validator.
> 
> I'm fine with including the changes here; in fact, since the PHP 8 tests _fail_ without them, then we need them before merging. I just didn't understand why you weren't using `in_array()` directly, and there were no comments in the code to indicate why.
> 
> Speaking of: it may be good to create a bug on bugs.php.net about this. The PHP 8 changelog does not mention anything about the change in behavior, and I don't see any bugs currently filed around it.

You're right, I did not explain why I removed the code, especially why I did not use the `in_array()` function.
Would you like me to fix InArray validator class adding why I don't use the `in_array()` in this PR?
It would be the same before reverting the requested changes.

weierophinney commented:
> Would you like me to fix InArray validator class adding why I don't use the `in_array()` in this PR?
> It would be the same before reverting the requested changes.

Yes, please!

jfernancordova committed:
Modifying InArray validator for PHP 8.0 compatibilities

Signed-off-by: José Fernando Cordova <jfernancordova@gmail.com>

jfernancordova committed:
Modifying InArray validator for PHP 8.0 compatibilities

Signed-off-by: José Fernando Cordova <jfernancordova@gmail.com>

jfernancordova commented:
Hi again @weierophinney! I started to modify the inArray validator, I realized that I did not need some functions I added, just changed the way of conditionals and using `in_array()` function, I ask you again if you want to explain this way but I think it is implicit.

weierophinney committed:
qa: remove redundant condition

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented on src/InArray.php:
```diff
@@ -216,9 +220,17 @@ public function isValid($value)
             }
 
-            if (in_array($value, $haystack, self::COMPARE_STRICT == $this->strict)) {
+            if (self::COMPARE_NOT_STRICT == $this->strict) {
                 return true;
             }
```
This part (and the related comparison in the recursive comparison section)... seems wrong. You're testing to see if a constant value referencing "not strict comparisons" matches the  flag provided by the consumer as to whether or not to do strict comparisons. You're not comparing the value **at all**, and yet if these match, you're returning true.

At best, this is only **accidentally** causing tests to pass. At worse, it's hiding something else.

So, I did some digging, and had quite a surprise: The behavior of `==` appears to have changed in PHP 8.

I did the following:

```php
$values = ['test', 0, 'A', 0.0, 'foo', 1, 'a', 'c'];
foreach ($values as $value) {
    printf("- Is b equal to %s?\t%s\n", $value, 'b' == $value ? 'Yes' : 'No');
}
```

Under PHP 7 versions, you get the following output:

```text
- Is b equal to test?	No
- Is b equal to 0?	Yes
- Is b equal to A?	No
- Is b equal to 0?	Yes
- Is b equal to foo?	No
- Is b equal to 1?	No
- Is b equal to a?	No
- Is b equal to c?	No
```

However, under PHP 8, we get:

```text
- Is b equal to test?	No
- Is b equal to 0?	No
- Is b equal to A?	No
- Is b equal to 0?	No
- Is b equal to foo?	No
- Is b equal to 1?	No
- Is b equal to a?	No
- Is b equal to c?	No
```

This is true even with strict_types off. And, obviously, this is hugely problematic for this particular comparison.

Looking in the [PHP 8 upgrade guide](https://github.com/php/php-src/blob/PHP-8.0/UPGRADING), there is a note about this:

> Non-strict comparisons between numbers and non-numeric strings now work by casting the number to string and comparing the strings. Comparisons between numbers and numeric strings continue to work as before. Notably, this means that `0 == "not-a-number"` is considered false now.

As such, I'm going to update the tests to vary the assertions based on PHP version, and I'll push the changes back to your branch.

jfernancordova replied:
Oh I see, I'm sorry for not realizing about this, I just trusted in the tests, thank you for the details, very appreciate @weierophinney

weierophinney committed:
fix: refactor and improve logic for usage under PHP 8

As it turns out, in PHP 8 when comparing a number and a non-numeric string, the engine casts the number to a string and then compares the strings.
This means that `'b' == 0` or `'1a' == 1` will now fail under PHP 8.
As such, I've updated some assertions to vary based on whether we are testing under PHP 8 and above, or PHP 7.
This also afforded me the opportunity to refactor the `InArray` logic to be more readable, as well as more correct.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

jfernancordova commented:
It was an amazing experience! Thanks for everything @boesing @glensc @weierophinney @Ocramius

lahivee commented:
thank you for your work on PHP8 support! when do you plan to release this?

weierophinney commented:
There's a few other PRs marked for the 2.14.0 release that I'm reviewing and helping complete now; hopefully later today, but it may be tomorrow.
